proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,484 @@
+a cbuffer or bound as a resource. So, if `other` is set, an error will always be emitted.
+
+These flags are all mutually exclusive, and are represented in the code with an enum.
+The `udt` flag has an associated flag, `contains_numeric`, which will be set if the
+UDT contains a numeric type.
+
+The next group of flags are the resource class flags, which are `srv`, `uav`, `cbv`,
+or `sampler`. For example, if an SRV resource is detected in the decl, then the 
+`resource` flag is set and the associated resource class flag `srv` is set too. For
+UDTs, the `udt` flag is set, and any resource class flags can be set if the associated
+resource is found contained within the UDT.
+
+The `default_globals` flag indicates whether or not the value ends up inside the 
+`$Globals` constant buffer. It will not be set if the decl appears inside a cbuffer 
+or tbuffer. The `$Globals` constant buffer will only be filled with non-HLSL-Objects.
+
+The `is_member` flag is set if the decl is a member declaration of a struct or class.
+This flag will inevitably lead to an error, because register annotations aren't allowed
+on member declarations. `err_hlsl_register_annotation_on_member` will be emitted.
+
+```
+Flags:
+  resource,
+  udt,
+  other,
+  basic,
+
+  srv,
+  uav,
+  cbv,
+  sampler
+
+  contains_numeric
+  default_globals
+  is_member
+
+struct Foo {
+  RWBuffer<int> rwbuf : register(u3); // is_member is set for this decl
+};
+
+RWBuffer<int> r0 : register(u0); // resource (uav is set)
+Foo f0 : register(u0); // udt (Foo doesn't contain a numeric, so contains_numeric isn't set)
+RayQuery<0> r1: register(t0); // other (error)
+float f1 : register (c0); // basic (ends up in $Globals constant buffer, so default_globals is set)
+```
+
+The first step is to simply check if the decl is inside a cbuffer or tbuffer
+block. If it isn't, and it has a numeric type, set the `default_globals` flag,
+because the value is bound to be placed in the `$Globals` buffer. If the decl is
+in a class or struct, it is a member declaration, and register annotations aren't
+allowed here, so we must set the `is_member` flag.
+
+Next, determine if the decl is a `cbuffer` or `tbuffer`. These are special in that they have
+their own decl class type, `HLSLBufferDecl`. To determine if the Decl is a `HLSLBufferDecl`,
+the compiler dynamically casts the Decl to an `HLSLBufferDecl`, and then if it succeeds,
+checks isCBuffer(). If true, the Decl is a cbuffer, otherwise it is a tbuffer. In either case,
+we know the decl is a `resource` and can infer the resource class (`cbv` or `srv`
+respectively). Otherwise, if the dynamic cast fails, we cast to `VarDecl`, and check 
+to see if the ResourceAttr attribute is present. If it is, the `resource` flag can be 
+set, and the corresponding resource class flag will be set. If the `VarDecl` dynamic cast
+fails, then there is some critical validation error, since any Decl with the register 

```

Comment:
>If the `VarDecl` dynamic cast
fails, then there is some critical validation error
>
What is the fault taulerance strategy here? Are you intending to have a default case to handle this?

Reply:
I was intending to check if either decl is castable, and if not an llvm_unreachable is executed, "unknown decl type"